packaging/macos/utility.m

```diff
@@ -0,0 +1,218 @@
+	if (!_mono_main)
+	{
+		NSLog(@"Could not load mono_main(): %s\n", dlerror());
+		return 1;
+	}
+
+	// Insert hostpath, dll and modId as arguments. Overwrite the first argument which was used to launch this application.
+	for (int i = argc; i > 0; i--)
+	{
+		argv[i + 1] = argv[i - 1];
+	}
+
+	argv[0] = (char*)[hostPath UTF8String];
+	argv[1] = (char*)[dllPath UTF8String];
+	argv[2] = modId;
+	argc += 2;
+
+	return _mono_main(argc, argv);
+}
+
+int launch_dotnet(int argc, char **argv, bool arm)
+{
+	NSString *dllPath;
+	NSString *hostPath;
+
+	if (arm)
+	{
+		hostPath = [exePath stringByAppendingPathComponent: @"arm64/libhostfxr.dylib"];;
+		dllPath = [exePath stringByAppendingPathComponent: @"arm64/OpenRA.Utility.dll"];
+	}
+	else
+	{
+		hostPath = [exePath stringByAppendingPathComponent: @"x86_64/libhostfxr.dylib"];;
+		dllPath = [exePath stringByAppendingPathComponent: @"x86_64/OpenRA.Utility.dll"];
+	}
+
+	void *lib = dlopen([hostPath UTF8String], RTLD_LAZY);
+	if (lib == NULL)
+	{
+		NSLog(@"Failed to load %@: %s\n", hostPath, dlerror());
+		return 1;
+	}
+
+	hostfxr_initialize_for_dotnet_command_line_fn hostfxr_initialize_for_dotnet_command_line = (hostfxr_initialize_for_dotnet_command_line_fn)dlsym(lib, "hostfxr_initialize_for_dotnet_command_line");
+	if (!hostfxr_initialize_for_dotnet_command_line)
+	{
+		NSLog(@"Could not load hostfxr_initialize_for_dotnet_command_line(): %s\n", dlerror());
+		return 1;
+	}
+
+	hostfxr_run_app_fn hostfxr_run_app = (hostfxr_run_app_fn)dlsym(lib, "hostfxr_run_app");
+	if (!hostfxr_run_app)
+	{
+		NSLog(@"Could not load hostfxr_run_app(): %s\n", dlerror());
+		return 1;
+	}
+
+	hostfxr_close_fn hostfxr_close = (hostfxr_close_fn)dlsym(lib, "hostfxr_close");
+	if (!hostfxr_close)
+	{
+		NSLog(@"Could not load hostfxr_close(): %s\n", dlerror());
+		return 1;
+	}
+
+	struct hostfxr_initialize_parameters params;
+	params.size = sizeof(params);
+	params.host_path = (char*)[[exePath stringByAppendingPathComponent: @"Utility"] UTF8String];
+	params.dotnet_root = dirname((char*)[hostPath UTF8String]);
+
+	// Insert dll and modId as arguments. Overwrite the first argument which was used to launch this application.
+	for (int i = argc; i > 0; i--)
+	{
+		argv[i] = argv[i - 1];
+	}
+
+	argv[0] = (char*)[dllPath UTF8String];
+	argv[1] = modId;
+	argc += 1;
+
+	hostfxr_handle host_context_handle;
+	hostfxr_initialize_for_dotnet_command_line(
+		argc,
+		argv,
+		&params,
+		&host_context_handle);
+
+	hostfxr_run_app(host_context_handle);
+
+	return hostfxr_close(host_context_handle);
+}
+
+int main(int argc, char **argv)
+{
+	BOOL useMono = NO;
+
+	if (@available(macOS 10.15, *))
+		useMono = [[[NSProcessInfo processInfo] environment]objectForKey:@"OPENRA_PREFER_MONO"] != nil;
+	else
+		useMono = YES;
+
+	size_t size;
+	cpu_type_t type;
+	size = sizeof(type);
+	bool arm = sysctlbyname("hw.cputype", &type, &size, NULL, 0) == 0 && (type & 0xFF) == CPU_TYPE_ARM;
+
+	if (useMono)
+	{
+		// mono is compiled for intel only
```

Comment:
This is a little awkward. Can remove a level of indentation by changing the above to something like:

```objective-c
BOOL useMono = NO;

cpu_type_t type;
size_t size = sizeof(type);
bool isArmArchitecture = sysctlbyname("hw.cputype", &type, &size, NULL, 0) == 0 && (type & 0xFF) == CPU_TYPE_ARM;

if (@available(macOS 10.15, *))
	useMono = !isArmArchitecture && [[[NSProcessInfo processInfo] environment]objectForKey:@"OPENRA_PREFER_MONO"] != nil;
else
	useMono = YES;
```

(macOS <= 10.15 did not support arm)

Reply:
polished up a bit differently